File: isValid.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{

    public:
         
         //Function to check for a valid parenthesis string
         bool isValidParenthesis(string s){
              //size of the string s
              int n = s.length();

              //Declare a stack 
              stack<int>st;

              //Iterate over the string
              for(int i=0;i<n;i++){
                 if(s[i] == '{' || s[i] == '[' || s[i] == '('){
                     st.push(s[i]);
                 }
                 else {
                  if(st.empty() or (s[i] != ')' and st.top() == '(') or (s[i] != ']' and st.top() == '[') or (st.top() == '{' and s[i] != '}'))
                      return false;
                   
                   st.pop();
                 }
              }

              return st.empty(); 
         }
}

File: isValid.py
class Solution:

    # Function to check for a valid parenthesis string
    def isValidParenthesis(self, s: str) -> bool:
        # size of the string s
        n = len(s)

        # Declare a stack
        st = []

        # Iterate over the string
        for i in range(n):
            if s[i] == '{' or s[i] == '[' or s[i] == '(':
                st.append(s[i])
            else:
                if not st or (s[i] != ')' and st[-1] == '(') or (s[i] != ']' and st[-1] == '[') or (st[-1] == '{' and s[i] != '}'):
                    return False
                st.pop()

        return not st
